### PR TITLE
Add `wholeMatch` and `prefixMatch`

### PR DESCRIPTION
**Pull request by itingliu**
Add the functions to string processing algorithms proposal and implement the
change.

Move the functions from `String` and `SubString` extensions to
`BidirectionalCollection`. Add tests for `firstMatch`, `wholeMatch`, and
`prefixMatch` that use a custom `BidirectionalCollection` type.

### CODE REVIEW

**Comment by milseman** [comment]
I agree that if these are on String/Substring, they should instead be generic like the others. I weakly hold that they should exist, but I appreciate @natecook1000's perspective, especially if we have anchoring modifier APIs on `Regex`. Let's discuss this at the next sync up.

**Comment by itingliu** [comment]
@swift-ci please test

File: Documentation/Evolution/StringProcessingAlgorithms.md [hunk 1]
```diff
@@ -473,7 +484,7 @@ extension RangeReplaceableCollection where SubSequence == Substring {
     /// - Returns: A new collection in which all occurrences of subsequence
     /// matching `regex` in `subrange` are replaced by `replacement`.
     public func replacing<R: RegexComponent, Replacement: Collection>(
-        _ regex: R,
+        _ r: R,
```
Comment (milseman) [inline Documentation/Evolution/StringProcessingAlgorithms.md:487]: Why this change? It makes the doc comment above invalid

File: Sources/_StringProcessing/Regex/Match.swift [hunk 2]
```diff
@@ -159,32 +159,23 @@ extension Regex {
 }
 
 @available(SwiftStdlib 5.7, *)
-extension String {
+extension BidirectionalCollection where SubSequence == Substring {
+  /// Match a regex in its entirety.
+  /// - Parameter r: The regex to match against.
```
Comment (milseman) [inline Sources/_StringProcessing/Regex/Match.swift:164]: Hmm... potential parameter names could be `regex` or `pattern` or `regexComponent`....

WDYT?

**Review by milseman — APPROVED** [review]
LGTM, but potentially different parameter names for docs.

File: Sources/_StringProcessing/Regex/Match.swift (hunk 2 above)
Comment (milseman) [reply Sources/_StringProcessing/Regex/Match.swift:164]: @natecook1000
